evennia/contrib/puzzles.py

```diff
@@ -0,0 +1,472 @@
+            return
+
+        caller.msg(
+            "Puzzle Recipe %s(%s) '%s' found.\nSpawning %d parts ..." % (
+            puzzle.name, puzzle.dbref, puzzle.db.puzzle_name, len(puzzle.db.parts)))
+
+        for proto_part in puzzle.db.parts:
+            # caller.msg('Protopart %r %r' % (proto_part, type(proto_part)))
+            part = spawn(proto_part)[0]
+            caller.msg("Part %s(%s) spawned and placed at %s(%s)" % (part.name, part.dbref, part.location, part.location.dbref))
+            part.mark_as_puzzle_member(puzzle.db.puzzle_name)
+
+        caller.msg("Puzzle armed |gsuccessfully|n.")
+
+
+class CmdUsePuzzleParts(MuxCommand):
+    """
+    Searches for all puzzles whose parts
+    match the given set of objects. If
+    there are matching puzzles, the result
+    objects are spawned in their corresponding
+    location if all parts have been passed in.
+
+    Usage:
+        use <part1[,part2,...>]
+    """
+
+    # TODO: consider allowing builder to provide
+    # messages and "hooks" that can be displayed
+    # and/or fired whenever the resolver of the puzzle
+    # enters the location where a result was spawned
+
+    key = 'use'
+    aliases = 'combine'
+    locks = 'cmd:pperm(use) or pperm(Player)'
+    help_category = 'Puzzles'
+
+    def func(self):
+        caller = self.caller
+
+        if not self.lhs:
+            caller.msg('Use what?')
+            return
+
+        many = 'these' if len(self.lhslist) > 1 else 'this'
+
+        # either all are parts, or abort finding matching puzzles
+        parts = []
+        partnames = self.lhslist[:]
+        for partname in partnames:
+            part = caller.search(
+                partname,
+                multimatch_string='Which %s. There are many.\n' % (partname),
+                nofound_string='There is no %s around.' % (partname)
+            )
+
+            if not part:
+                return
+
+            if not part.tags.get(_PUZZLES_TAG_MEMBER, category=_PUZZLES_TAG_CATEGORY) \
+                or not inherits_from(part, PuzzlePartObject):
+
+                # not a puzzle part ... abort
+                caller.msg('You have no idea how %s can be used' % (many))
+                return
+
+            # a valid part
+            parts.append(part)
+
+        # Create lookup dict
+        parts_dict = dict((part.dbref, part) for part in parts)
+
+        # Group parts by their puzzle name
+        puzzle_ingredients = dict()
+        for part in parts:
+            puzzle_name = part.db.puzzle_name
+            if puzzle_name not in puzzle_ingredients:
+                puzzle_ingredients[puzzle_name] = []
+            puzzle_ingredients[puzzle_name].append(
+                (part.dbref, proto_def(part, with_tags=False))
+            )
+
+        # Find all puzzles by puzzle name
+        # FIXME: we rely on obj.db.puzzle_name which is visible and may be cnaged afterwards. Can we lock it and hide it?
```

Comment:
@Henddher As mentioned, if you use tag-names in your recipe (instead of, as now, full objects), you could use the tag's category to identify it as a component of a particular puzzle or just as being a generic puzzle piece at all. As long as the result of the puzzle is also tagged as a puzzle-piece you could then use that in another puzzle as much as you wanted. 

So the creation of a recipie would then be a lot easier (example of syntax)

```
@puzzle smoothie: orange + mango + yoghurt + blender = fruit smoothie;smoothie:typeclasses.objects.Smoothie
```
This creates a new Recipie (a Script) named `smoothie` holding the recipie tag-names) and the instructions for how to build the resulting object. You could now use `@armrecipe` to build these components or manually build them and tag each with a Tag `orange` and category `recipe_ingredient` or something. Then the `orange` component could actually be named (have a `key`) `An yellow-red fruit` or `A rotting orange` or "Peeled orange" to make the puzzle harder or to match your game world. It also means that multiple objects could act as components in the recipe if you wanted to, as long as they have the right tag. "A can of preserved orange slices" works just as well as "A big orange" as long as they are both tagged right. 

All you now need to do is make sure the resulting created object is created with a matching tag. Maybe you'll default to doing so (so give the `fruit smoothie` a Tag `fruit smoothie` always). You can then naturally use this with another recipe: 

```
@recipe fruit_grenade:fruit smoothie + exploding bananas + metal can = fruit grenade:typeclasses.objects.FruitGrenade
```

Since all you specify for your recipe are strings, it becomes very easy to set up and your entire code becomes a lot simpler. Of course, here I suggested specifying typeclass for the resulting object since I expect it to be created on the fly when the recipe is completed. You could also give a prototype if you preferred.